src/wallet/wallet.cpp

```diff
@@ -1059,7 +1059,19 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, CWalletTx::Co
             for (const CTxOut& txout: tx.vout) {
                 const auto& spk_man = GetScriptPubKeyMan(txout.scriptPubKey);
                 if (!spk_man) continue;
-                spk_man->MarkUnusedAddresses(txout.scriptPubKey);
+                for (auto& dest : spk_man->MarkUnusedAddresses(txout.scriptPubKey)) {
+                    // If internal flag is not defined try to infer it from the ScriptPubKeyMan
+                    if (!dest.internal.has_value()) {
+                        dest.internal = IsInternalScriptPubKeyMan(spk_man);
+                    }
+
+                    // If this is a receiving address and it's not in the address book yet
+                    // (e.g. it wasn't generated on this node or we're restoring from backup)
+                    // add it to the address book for proper transaction accounting
+                    if (!dest.internal.value_or(true) && !FindAddressBookEntry(dest.dest, /* allow_change= */ false)) {
```

Comment:
In cbb90fbc00e345100587cd47926c6c7fc5487628 "Automatically add labels to detected receiving addresses"

Shouldn't we be marking everything that doesn't have `dest.internal` as a receiving address? i.e.

```suggestion
                   if (!dest.internal.value_or(false) && !FindAddressBookEntry(dest.dest, /* allow_change= */ false)) {
```

Reply:
I want to be conservative and skip if we can't determine if that's an change or receiving destination. But the code is not obvious and I restructured it to make it explicit.